packages/protocol/src/protocol/ExchangeManager.ts

```diff
@@ -116,31 +122,42 @@ export class ExchangeManager {
         if (this.#isClosing) {
             return;
         }
+
+        using closing = this.#lifetime.closing();
+
         this.#isClosing = true;
 
+        const exchangesClosed = new BasicMultiplex();
+
         for (const exchange of this.#exchanges.values()) {
-            this.#workers.add(exchange.close(true), `closing exchange ${exchange.via}`);
+            exchangesClosed.add(exchange.close(true));
         }
 
-        await this.#workers;
-
-        for (const listeners of this.#listeners.keys()) {
-            this.#deleteTransport(listeners);
+        {
+            using _closing = closing.join("exchanges");
+            await exchangesClosed;
         }
 
-        await this.#workers;
+        for (const listener of this.#listeners.keys()) {
+            this.#deleteTransport(listener);
+        }
 
         for (const protocol of this.#protocols.values()) {
-            await protocol.close();
+            this.#workers.add(protocol.close());
         }
 
-        await this.#workers;
+        {
```

Comment:
Ohhh yeeaaahh ... such intermediate blocks might become handy ... but also strange ... i think I need a bit to get familiar with that. :-)